File: skills/sn-setup/templates/claude/hooks/chokepoint-gate.py
"""PreToolUse chokepoint-gate hook for the Claude Agent SDK (Python).

Reads `.harness/chokepoints.yaml` and blocks Edit/Write/etc. calls targeting
any listed file or glob. Use to keep agents off load-bearing files until a
human OKs the change.
"""
from __future__ import annotations

import fnmatch
import re
from pathlib import Path
from typing import Any


def _project_root(start: Path | None = None) -> Path:
    p = (start or Path.cwd()).resolve()
    while p != p.parent:
        if (p / ".sn-init").exists() or (p / "CLAUDE.md").exists():
            return p
        p = p.parent
    return Path.cwd().resolve()


def _patterns() -> list[str]:
    root = _project_root()
    path = root / ".harness" / "chokepoints.yaml"
    if not path.exists():
        return []
    items: list[str] = []
    in_block = False
    for line in path.read_text(encoding="utf-8").splitlines():
        if re.match(r"^\s*chokepoints\s*:\s*$", line):
            in_block = True
            continue
        if in_block:
            m = re.match(r"^\s*-\s*(.+?)\s*$", line)
            if m:
                items.append(m.group(1))
            elif line.strip() and not line.lstrip().startswith("#"):
                # next top-level key terminates the block
                break
    return items


def _target_path(input_data: dict) -> str | None:
    ti = (input_data or {}).get("tool_input") or {}
    for key in ("file_path", "path", "target"):
        val = ti.get(key)
        if isinstance(val, str) and val:
            return val
    return None


async def chokepoint_gate_hook(input_data: dict, tool_use_id: str | None, context: Any) -> dict:
    target = _target_path(input_data)
    if target is None:
        return {}

    root = _project_root()
    try:
        relative = str(Path(target).resolve().relative_to(root))
    except ValueError:
        relative = target

    for pat in _patterns():
        if fnmatch.fnmatch(relative, pat):
            return {
                "action": "block",
                "reason": (
                    f"chokepoint-gate: '{relative}' matches '{pat}'. "
                    "Edit this file only after a human approves the change."
                ),
            }
    return {}

File: platform-marketplace/plugins/core-guardrails/hooks/chokepoint-gate.ts
// PreToolUse chokepoint-gate hook for the Claude Agent SDK (TypeScript).
//
// Reads `.harness/chokepoints.yaml` and blocks Edit/Write/etc. calls
// targeting any listed file or glob.

import { readFile } from "node:fs/promises";
import { existsSync } from "node:fs";
import { dirname, join, relative as relPath, resolve } from "node:path";

function projectRoot(start: string = process.cwd()): string {
  let p = resolve(start);
  while (p !== dirname(p)) {
    if (existsSync(join(p, ".sn-init")) || existsSync(join(p, "CLAUDE.md"))) {
      return p;
    }
    p = dirname(p);
  }
  return resolve(start);
}

async function patterns(): Promise<string[]> {
  const root = projectRoot();
  const path = join(root, ".harness", "chokepoints.yaml");
  if (!existsSync(path)) return [];
  const text = await readFile(path, "utf-8");
  const lines = text.split(/\r?\n/);
  const items: string[] = [];
  let inBlock = false;
  for (const ln of lines) {
    if (/^\s*chokepoints\s*:\s*$/.test(ln)) {
      inBlock = true;
      continue;
    }
    if (inBlock) {
      const m = ln.match(/^\s*-\s*(.+?)\s*$/);
      if (m) {
        items.push(m[1]);
      } else if (ln.trim() && !ln.trim().startsWith("#")) {
        break;
      }
    }
  }
  return items;
}

function targetPath(input: any): string | null {
  const ti = input?.tool_input ?? {};
  for (const k of ["file_path", "path", "target"]) {
    if (typeof ti[k] === "string" && ti[k]) return ti[k];
  }
  return null;
}

function globToRegExp(glob: string): RegExp {
  // Minimal: ** matches anything, * matches single segment.
  const re = glob
    .split("**").map((seg) => seg.replace(/[.+^${}()|[\]\\]/g, "\\$&").replace(/\*/g, "[^/]*"))
    .join(".*");
  return new RegExp("^" + re + "$");
}

export async function chokepointGateHook(input: any, _toolUseId: string | undefined, _ctx: any): Promise<Record<string, unknown>> {
  const t = targetPath(input);
  if (!t) return {};
  const root = projectRoot();
  let relative: string;
  try {
    relative = relPath(root, resolve(t));
  } catch {
    relative = t;
  }
  for (const pat of await patterns()) {
    const re = globToRegExp(pat);
    if (re.test(relative)) {
      return {
        action: "block",
        reason: `chokepoint-gate: '${relative}' matches '${pat}'. Edit this file only after a human approves the change.`,
      };
    }
  }
  return {};
}
